Firestore/core/src/firebase/firestore/local/leveldb_key.cc

```diff
@@ -108,300 +109,276 @@ enum ComponentLabel {
+      absl::string_view tmp = MakeStringView(src_);
+      if (OrderedCode::ReadString(&tmp, &result)) {
+        src_ = MakeSlice(tmp);
+        return result;
+      }
+    }
 
-/**
- * For each segment in the given resource path writes an
- * ComponentLabel::PathSegment component label and a string containing the path
- * segment.
- */
-void WriteResourcePath(std::string *dest, const ResourcePath &path) {
-  for (const auto &segment : path) {
-    WriteComponentLabel(dest, ComponentLabel::PathSegment);
-    OrderedCode::WriteString(dest, segment);
+    Fail();
+    return "";
   }
-}
 
-/**
- * Reads component labels and strings from the given key contents until it finds
- * a component label other that ComponentLabel::PathSegment. All matched path
- * segments are assembled into a resource path and wrapped in an DocumentKey.
- *
- * If the read is unsuccessful or the document key is invalid, returns false,
- * and changes none of its arguments.
- *
- * If the read is successful, returns true, contents will be updated to the next
- * unread byte, and value will be set to the decoded document key.
- */
-bool ReadDocumentKey(leveldb::Slice *contents, DocumentKey *result) {
-  leveldb::Slice complete_segments = *contents;
-
-  std::string segment;
-  std::vector<std::string> path_segments;
-  for (;;) {
-    // Advance a temporary slice to avoid advancing contents into the next key
-    // component which may not be a path segment.
-    leveldb::Slice read_position = complete_segments;
-    if (!ReadComponentLabelMatching(&read_position,
-                                    ComponentLabel::PathSegment)) {
-      break;
-    }
-    if (!ReadString(&read_position, &segment)) {
-      return false;
+  /**
+   * Reads a component label from the key.
+   *
+   * If the read is unsuccessful, returns ComponentLabel::Unknown and fails the
+   * Reader.
+   *
+   * Otherwise, returns the ComponentLabel and advances the Reader to the next
+   * unread byte.
+   */
+  ComponentLabel ReadComponentLabel() {
+    if (ok_) {
+      int64_t raw_result = ReadSignedNumIncreasing();
+      if (ok_ && raw_result >= ComponentLabel::Terminator &&
+          raw_result <= ComponentLabel::Unknown) {
+        return static_cast<ComponentLabel>(raw_result);
+      }
     }
 
-    path_segments.push_back(std::move(segment));
-    segment.clear();
-
-    complete_segments = read_position;
+    Fail();
+    return ComponentLabel::Unknown;
   }
 
-  ResourcePath path{std::move(path_segments)};
-  if (path.size() > 0 && DocumentKey::IsDocumentKey(path)) {
-    *contents = complete_segments;
-    *result = DocumentKey{std::move(path)};
-    return true;
-  }
+  /**
+   * Reads a component label from the key.
+   *
+   * If the read is unsuccessful, returns false, and fails the Reader.
+   *
+   * Otherwise returns whether or not the component label is equal to the
+   * `expected_label` and advances the Rader to the next unread byte.
+   */
+  ABSL_MUST_USE_RESULT
```

Comment:
This looks interesting.

Reply:
I missed one in my early implementation and this caused me to go find this.